lib/semantics/resolve-names.cc

```diff
@@ -1037,11 +1037,271 @@ class ConstructVisitor : public DeclarationVisitor {
+    // common block
+    if (auto *symbol{ResolveOmpCommonBlockName(name)}) {
+      CheckMultipleAppearances(*name, symbol, Symbol::Flag::OmpCommonBlock);
+      // 2.15.3 When a named common block appears in a list, it has the same
+      // meaning as if every explicit member of the common block appeared in
+      // the list
+      for (Symbol *object : symbol->get<CommonBlockDetails>().objects()) {
+        ResolveOmp(*object, ompFlag);
+      }
+    } else {
+      Say(name->source,  // 2.15.3
+          "COMMON block must be declared in the same scoping unit "
+          "in which the OpenMP directive or clause appears"_err_en_US);
+    }
+  }
+}
+
+Symbol *OmpVisitor::ResolveOmp(const parser::Name &name, Symbol::Flag ompFlag) {
+  if (ompFlagsRequireNewSymbol.test(ompFlag)) {
+    return DeclarePrivateAccessEntity(name, ompFlag);
+  } else {
+    return DeclareOrMarkOtherAccessEntity(name, ompFlag);
+  }
+}
+
+Symbol *OmpVisitor::ResolveOmp(Symbol &symbol, Symbol::Flag ompFlag) {
+  if (ompFlagsRequireNewSymbol.test(ompFlag)) {
+    return DeclarePrivateAccessEntity(symbol, ompFlag);
+  } else {
+    return DeclareOrMarkOtherAccessEntity(symbol, ompFlag);
+  }
+}
+
+Symbol *OmpVisitor::DeclarePrivateAccessEntity(
+    const parser::Name &name, Symbol::Flag ompFlag) {
+  Symbol &prev{FindOrDeclareEnclosingEntity(name)};
+  if (prev.owner() != currScope()) {
+    auto &symbol{MakeSymbol(name, HostAssocDetails{prev})};
+    symbol.set(ompFlag);
+    name.symbol = &symbol;  // override resolution to parent
+    return &symbol;
+  } else {
+    prev.set(ompFlag);
+    return &prev;
+  }
+}
+
+Symbol *OmpVisitor::DeclarePrivateAccessEntity(
+    Symbol &object, Symbol::Flag ompFlag) {
+  if (object.owner() != currScope() &&
+      !FindInScope(currScope(), object.name())) {
+    auto &symbol{MakeSymbol(object.name(), Attrs{}, HostAssocDetails{object})};
+    symbol.set(ompFlag);
+    return &symbol;
+  } else {
+    object.set(ompFlag);
+    return &object;
+  }
+}
+
+Symbol *OmpVisitor::DeclareOrMarkOtherAccessEntity(
+    const parser::Name &name, Symbol::Flag ompFlag) {
+  Symbol &prev{FindOrDeclareEnclosingEntity(name)};
+  name.symbol = &prev;
+  if (ompFlagsRequireMark.test(ompFlag)) {
+    prev.set(ompFlag);
+  }
+  return &prev;
+}
+
+Symbol *OmpVisitor::DeclareOrMarkOtherAccessEntity(
+    Symbol &object, Symbol::Flag ompFlag) {
+  if (ompFlagsRequireMark.test(ompFlag)) {
+    object.set(ompFlag);
+  }
+  return &object;
+}
+
+static bool WithMultipleAppearancesException(
+    const Symbol *symbol, Symbol::Flag ompFlag) {
+  return (ompFlag == Symbol::Flag::OmpFirstPrivate &&
+             symbol->test(Symbol::Flag::OmpLastPrivate)) ||
+      (ompFlag == Symbol::Flag::OmpLastPrivate &&
+          symbol->test(Symbol::Flag::OmpFirstPrivate));
+}
+
+void OmpVisitor::CheckMultipleAppearances(
+    const parser::Name &name, const Symbol *symbol, Symbol::Flag ompFlag) {
```

Comment:
Can `symbol` ever be null?

Reply:
Answering the above two questions: no. `WithMultipleAppearancesException` will be called in this `CheckMultipleAppearances`. And the actual argument of the `const Symbol *symbol` here comes from the result of `ResolveOmp` which will either find the existing symbol (in the enclosing scope) or create a new symbol with `HostAssoc` detail.

Reply:
If the pointer cannot be null, then a reference is better than a pointer.

Reply:
Thanks! Tim just suggested the same, I will go through the PR and make the necessary changes.

Reply:
The necessary changes have been made.